.github/workflows/matrix.yml

```diff
@@ -199,28 +199,28 @@ jobs:
     - name: prepare
       if: ${{ env.SKIP == 'false' }}
       run: bash ./build-scripts/requirements.sh
-    - name: Get Date
-      id: get-date
+    - name: Get ccache vars
+      id: get-vars
       if: ${{ env.SKIP == 'false' }}
       run: |
         echo "::set-output name=date::$(/bin/date -u "+%Y%m%d%H%M")"
+        echo "::set-output name=ccache-path::$([ "$RUNNER_OS" = "macOS" ] && echo '/Users/runner/Library/Caches/ccache' || echo '~/.ccache')"
       shell: bash
-    - name: ccache cache files (ubuntu)
-      if: ${{ env.SKIP == 'false' && runner.os == 'Linux' }}
+    - name: ccache cache files (master)
+      if: ${{ github.event_name == 'push' && env.SKIP == 'false' && ( runner.os == 'Linux' || runner.os == 'macOS' ) }}
       uses: actions/cache@v2
       with:
-        path: ~/.ccache
-        key: ccache-linux-${{ matrix.compiler }}-${{ steps.get-date.outputs.date }}
+        path: ${{ steps.get-vars.outputs.ccache-path }}
+        # double-dash after compiler is not a typo, it is to disambiguate between g++-<date> and g++-11-<date> for restore key prefix matching
+        key: ccache-${{ runner.os }}-${{ matrix.compiler }}--${{ steps.get-vars.outputs.date }}
         restore-keys: |
-          ccache-linux-${{ matrix.compiler }}-
-    - name: ccache cache files (mac)
-      if: ${{ env.SKIP == 'false' && runner.os == 'macOS' }}
+          ccache-${{ runner.os }}-${{ matrix.compiler }}--
+    - name: ccache cache files (PR)
+      if: ${{ github.event_name == 'pull_request' && env.SKIP == 'false' && ( runner.os == 'Linux' || runner.os == 'macOS' ) }}
       uses: actions/cache@v2
       with:
-        path: /Users/runner/Library/Caches/ccache
-        key: ccache-mac-${{ matrix.compiler }}-${{ steps.get-date.outputs.date }}
-        restore-keys: |
-          ccache-mac-${{ matrix.compiler }}-
+        path: ${{ steps.get-vars.outputs.ccache-path }}
+        key: ccache-${{ runner.os }}-${{ matrix.compiler }}--
```

Comment:
My read of the documentation is that we need to list this explicitly in `restore-keys` to get prefix matching. The primary key is always exact match. https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows#matching-a-cache-key

Reply:
The cache is implicitly namespaced by branch, we're relying on cache entities emitted by master filling up the cache store and evicting everything else.
However if the main key doesn't do partial matches that's all for naught.

Reply:
Ok so we can calculate "todays" cache and use that as a key, which kinda sucks but not much to be done about that.

Reply:
Main cache would create fresh cache archives and would kick out older main cache archives, but because of the LRU eviction policy the undated PR cache archive would stay refreshed. I think having a 'today PR cache' is fine, using the restore key to get a hit off the main branch by prefix.

Reply:
ok lemme walk through this.
Initial state: cache store is full of junk from yesterday.
Then we have PR builds:
With no fallback they just get no cache and take a long time.
With a fallback, they will build faster, but will thrash the cache store by pushing new cache entries, most of which will never have hits.
Once there's a master build (possibly scheduled to queue right as the date ticks over)
With no PR fallback, it will pick up yesterdays cache with its own fallback, then push a new master entry.
With a PR fallback, it will miss, but still push a new master entry.
Then we're back to steady state where either way PRs are pulling todays cache entry.

Here's what I'm worried about though, we tend to have tons of builds in flight at the same time, so they're frequently concurrent.
master build starts and pulls fallback cache
PR build starts and pulls fallback cache
master build completes and pushes new cache entry.
PR build completes and pushes new cache entry, evicting a master cache entry.
Then we go an unbounded amount of time with no cache entries.

Reply:
That seems pretty unlikely, unless we're extremely strapped for cache space. The entries that are going to get evicted first are going to be all of the oldest master builds and yesterday's PR cache.

What about this below? Separate master and PR cache namespaces. Use date only for PR cache key, not datetime. PR cache always falls back to most recent master. I think the worst that can happen is master runs without a cache if we are running on the wire and at 12:01 a PR runs and uploads a cache evicting the master cache that it just picked up.

```
    - name: Get ccache vars
      id: get-vars
      if: ${{ env.SKIP == 'false' }}
      run: |
        echo "::set-output name=date::$(/bin/date -u "+%Y%m%d")"
        echo "::set-output name=datetime::$(/bin/date -u "+%Y%m%d%H%M")"
        echo "::set-output name=ccache-path::$([ "$RUNNER_OS" = "macOS" ] && echo '/Users/runner/Library/Caches/ccache' || echo '~/.ccache')"
      shell: bash
    - name: ccache cache files (master)
      if: ${{ github.event_name == 'push' && env.SKIP == 'false' && ( runner.os == 'Linux' || runner.os == 'macOS' ) }}
      uses: actions/cache@v2
      with:
        path: ${{ steps.get-vars.outputs.ccache-path }}
        # double-dash after compiler is not a typo, it is to disambiguate between g++-<date> and g++-11-<date> for restore key prefix matching
        # ccache-master- namespace disambiguates from PR namespace
        key: ccache-master-${{ runner.os }}-${{ matrix.compiler }}--${{ steps.get-vars.outputs.datetime }}
        restore-keys: |
          ccache-master-${{ runner.os }}-${{ matrix.compiler }}--
    - name: ccache cache files (PR)
      if: ${{ github.event_name == 'pull_request' && env.SKIP == 'false' && ( runner.os == 'Linux' || runner.os == 'macOS' ) }}
      uses: actions/cache@v2
      with:
        path: ${{ steps.get-vars.outputs.ccache-path }}
        # ccache-pr- namespace for primary keys, using date only, not datetime
        key: ccache-pr-${{ runner.os }}-${{ matrix.compiler }}--${{ steps.get-vars.outputs.date }}
        restore-keys: |
          # restore keys using master namespace to pick up most recent master for fallback, not yesterday's PR cache.
          ccache-master-${{ runner.os }}-${{ matrix.compiler }}--
```

Reply:
> That seems pretty unlikely, unless we're extremely strapped for cache space. 

We are, we have 7 build configurations and about 9GB to play with, ideally each entry would be around 2GB, but obviously we're going to have to cut it down to 1.2GB max.